[Apple Pay] PKPaymentRequest instances are vended generic user agent string on iOS

aprotyas opened the pull request:
#### 6e4f73ab3bd2705006528afdfc8b17fedbfb96a2
<pre>
[Apple Pay] PKPaymentRequest instances are vended generic user agent string on iOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=271503">https://bugs.webkit.org/show_bug.cgi?id=271503</a>
<a href="https://rdar.apple.com/125267234">rdar://125267234</a>

Reviewed by Aditya Keerthi.

WebPaymentCoordinatorProxy either lives in the UI process (macOS) or in
the network process (iOS). In 270312@main, we unconditionally consulted
WebPageProxy::userAgent() when populating PKPaymentRequest instances
with originating user agent strings. This approach is not valid for iOS
because we would be trying to access a WebPageProxy (lives in the UI
process) directly from an object that lives in the network process. As a
result, we ended up supplying the fallback, generic UA strings on iOS.

This patch addresses this bug by instead delegating the work of
procuring a UA string to WebPaymentCoordinatorProxy::Client. On macOS,
the client can simply query WebPageProxy::userAgent() since it lives in
the UI process. On iOS, we add plumbing to be able to send an async
message from the network process across the UI process connection asking
for an appropriate UA string.

* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/ios/NetworkConnectionToWebProcessIOS.mm:
(WebKit::NetworkConnectionToWebProcess::getPaymentCoordinatorEmbeddingUserAgent):
* Source/WebKit/Shared/ApplePay/WebPaymentCoordinatorProxy.h:
* Source/WebKit/Shared/ApplePay/cocoa/WebPaymentCoordinatorProxyCocoa.mm:
(WebKit::WebPaymentCoordinatorProxy::platformSetPaymentRequestUserAgent):

* Source/WebKit/Shared/ApplePay/ios/WebPaymentCoordinatorProxyIOS.mm:
(WebKit::WebPaymentCoordinatorProxy::platformShowPaymentUI):
* Source/WebKit/Shared/ApplePay/mac/WebPaymentCoordinatorProxyMac.mm:
(WebKit::WebPaymentCoordinatorProxy::platformShowPaymentUI):

On both platforms, delegate work that needs to be performed with the
payment authorization presenter to the completion handler executed by
WebPaymentCoordinatorProxy::Client once it receives the user agent
string. We do so because the payment authorization presenter must only
be used after our PKPaymentRequest instance is fully fleshed out, which
includes having had its userAgent property assigned correctly.

* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::Internals::getPaymentCoordinatorEmbeddingUserAgent):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.messages.in:
* Source/WebKit/UIProcess/Network/NetworkProcessProxyCocoa.mm:
(WebKit::NetworkProcessProxy::getPaymentCoordinatorEmbeddingUserAgent):
* Source/WebKit/UIProcess/WebPageProxyInternals.h:

Canonical link: <a href="https://commits.webkit.org/276712@main">https://commits.webkit.org/276712@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ce0f8dbbe5c870f596f3d03a2867510c5a7353c0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45422 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24543 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47950 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/48087 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41430 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/28777 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21938 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/48087 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46000 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/21619 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/48087 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/19030 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [❌ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3468 "Failed to checkout and rebase branch from PR 26371") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/41719 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/40591 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49810 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/20406 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/16935 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/44299 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/21713 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/43125 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/22074 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6320 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/21401 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/Network/NetworkProcessProxy.messages.in:
```diff
@@ -85,6 +85,10 @@ messages -> NetworkProcessProxy LegacyReceiver {
 #endif
 
+#if ENABLE(APPLE_PAY_REMOTE_UI)
+    PaymentCoordinatorEmbeddingUserAgent(WebKit::WebPageProxyIdentifier webPageProxyIdentifier) -> (String userAgent) Synchronous
```
Does this have to be done synchronously? Looking at the code, it seems like it could be rewritten to make it async.

aprotyas replied:
It doesn't have to be synchronous. I'd gone with this because the asynchronous version makes the call site a little verbose, but on second thought that's a tough sell.

aprotyas replied:
(changing to async...)

aprotyas replied:
Done! https://github.com/WebKit/WebKit/commit/5f513a0a1ef25c698b67c93740b9a97ea9dcb822

aprotyas committed:
[Apple Pay] PKPaymentRequest instances are vended generic user agent string on iOS
https://bugs.webkit.org/show_bug.cgi?id=271503
rdar://125267234

Reviewed by Aditya Keerthi.

WebPaymentCoordinatorProxy either lives in the UI process (macOS) or in
the network process (iOS). In 270312@main, we unconditionally consulted
WebPageProxy::userAgent() when populating PKPaymentRequest instances
with originating user agent strings. This approach is not valid for iOS
because we would be trying to access a WebPageProxy (lives in the UI
process) directly from an object that lives in the network process. As a
result, we ended up supplying the fallback, generic UA strings on iOS.

This patch addresses this bug by instead delegating the work of
procuring a UA string to WebPaymentCoordinatorProxy::Client. On macOS,
the client can simply query WebPageProxy::userAgent() since it lives in
the UI process. On iOS, we add plumbing to be able to send an async
message from the network process across the UI process connection asking
for an appropriate UA string.

* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/ios/NetworkConnectionToWebProcessIOS.mm:
(WebKit::NetworkConnectionToWebProcess::getPaymentCoordinatorEmbeddingUserAgent):
* Source/WebKit/Shared/ApplePay/WebPaymentCoordinatorProxy.h:
* Source/WebKit/Shared/ApplePay/cocoa/WebPaymentCoordinatorProxyCocoa.mm:
(WebKit::WebPaymentCoordinatorProxy::platformSetPaymentRequestUserAgent):

* Source/WebKit/Shared/ApplePay/ios/WebPaymentCoordinatorProxyIOS.mm:
(WebKit::WebPaymentCoordinatorProxy::platformShowPaymentUI):
* Source/WebKit/Shared/ApplePay/mac/WebPaymentCoordinatorProxyMac.mm:
(WebKit::WebPaymentCoordinatorProxy::platformShowPaymentUI):

On both platforms, delegate work that needs to be performed with the
payment authorization presenter to the completion handler executed by
WebPaymentCoordinatorProxy::Client once it receives the user agent
string. We do so because the payment authorization presenter must only
be used after our PKPaymentRequest instance is fully fleshed out, which
includes having had its userAgent property assigned correctly.

* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::Internals::getPaymentCoordinatorEmbeddingUserAgent):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.messages.in:
* Source/WebKit/UIProcess/Network/NetworkProcessProxyCocoa.mm:
(WebKit::NetworkProcessProxy::getPaymentCoordinatorEmbeddingUserAgent):
* Source/WebKit/UIProcess/WebPageProxyInternals.h:

Canonical link: https://commits.webkit.org/276712@main
